packages/hdwallet-provider/src/index.ts

```diff
@@ -215,18 +225,94 @@ class HDWalletProvider {
     });
   }
 
-  public send(
+  private initialize(): Promise<void> {
+    return new Promise(resolve => {
+      this.sendAsync({
+        jsonrpc: '2.0',
+        id: Date.now(),
+        method: 'eth_chainId',
+        params: []
+      }, (_error, response) => {
+        if (response && response.result) {
+          this.chainId = response.result;
+        }
+        resolve();
+      });
+    });
+  }
+
+  // private helper to check if given mnemonic uses BIP39 passphrase protection
+  private checkBIP39Mnemonic ({
+    addressIndex,
+    numberOfAddresses,
+    phrase,
+    password
+  }: {
+    addressIndex: number;
+    numberOfAddresses: number;
+    phrase: string;
+    password?: string;
+  }) {
+    this.hdwallet = EthereumHDKey.fromMasterSeed(
+      bip39.mnemonicToSeedSync(phrase, password)
+    );
+
+    if (!bip39.validateMnemonic(phrase, wordlist)) {
+      throw new Error("Mnemonic invalid or undefined");
+    }
+
+    // crank the addresses out
+    for (let i = addressIndex; i < addressIndex + numberOfAddresses; i++) {
+      const wallet = this.hdwallet
+        .derivePath(this.walletHdpath + i)
+        .getWallet();
+      const addr = `0x${wallet.getAddress().toString("hex")}`;
+      this.addresses.push(addr);
+      this.wallets[addr] = wallet;
+    }
+  }
+
+  // private helper leveraging ethUtils to populate wallets/addresses
+  private ethUtilValidation ({
+    addressIndex,
+    privateKeys
+  }: {
+    addressIndex: number;
+    privateKeys: string[];
+  }) {
+    // crank the addresses out
+    for (let i = addressIndex; i < privateKeys.length; i++) {
+      const privateKey = Buffer.from(privateKeys[i].replace("0x", ""), "hex");
+      if (EthUtil.isValidPrivate(privateKey)) {
+        const wallet = ethJSWallet.fromPrivateKey(privateKey);
+        const address = wallet.getAddressString();
+        this.addresses.push(address);
+        this.wallets[address] = wallet;
+      }
+    }
+  };
+
+
+  public async send(
     payload: JSONRPCRequestPayload,
     callback: JSONRPCErrorCallback | Callback<JsonRPCResponse>
-  ): void {
+  ): Promise<void> {
+    await this.initialized;
     return this.engine.send.call(this.engine, payload, callback);
   }
 
   public sendAsync(
     payload: JSONRPCRequestPayload,
     callback: JSONRPCErrorCallback | Callback<JsonRPCResponse>
   ): void {
-    this.engine.sendAsync.call(this.engine, payload, callback);
+    // @ts-ignore
+    if (this.initialized) {
+      this.initialized.then(() => {
+        this.engine.sendAsync.call(this.engine, payload, callback);
```

Comment:
note: I don't think the `.call(this.engine` is necessary here, as the "call-site" is already `this.engine`.

Reply:
Maybe someone may have been passing the `sendAsync` function around (unattached from this.engine)? It might be nice to clean it up since I noticed it (verify nothing weird is happening first, of course).